add tests to assert on expected behavior of  wrapped ResponseWriter objects

tkashem opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
Add tests to verify expected behavior of the `ResponseWriter` object(s) wrapped by the handlers in the handler chain

#### Which issue(s) this PR fixes:
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/4462
```

tkashem committed:
move fake.go to testing package for reusability

tkashem committed:
add a test to verify the chain of wrapped ResponseWriter

add a test to walk through the chain of wrapped ResponseWriter
objects using the full handler chain of the generic apiserver
object, and ensure that interface compatibility is maintained

tkashem committed:
test each ResponseWriter decorator with net/http server

tkashem commented:
/cc @wojtek-t @sttts

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/webhook_duration_test.go:
```diff
@@ -0,0 +1,54 @@
+/*
+Copyright 2023 The Kubernetes Authors.
```
2024

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/webhook_duration_test.go:
```diff
@@ -0,0 +1,54 @@
+
+	decorator, ok := outer.(responsewriter.UserProvidedDecorator)
+	if !ok {
+		t.Errorf("expected the middle to implement UserProvidedDecorator")
```
t.Fatal

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/webhook_duration_test.go:
```diff
@@ -0,0 +1,54 @@
+	// http.CloseNotifier, or http.Hijacker; so WrapForHTTP1Or2 is not
+	// expected to return an outer object.
+	if outer != middle {
+		t.Errorf("did not expect a new outer object, but got %v", outer)
```
t.Fatal?

tkashem replied:
the test can still continue to the next assertion, so using `Errorf` here, thoughts?

sttts replied:
oh, then it's fine. Wasn't sure about that.

fedebongio commented:
/triage accepted

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/testing/fake.go:
```diff
@@ -89,3 +89,26 @@ func AssertResponseWriterInterfaceCompatibility(t *testing.T, inner, outer http.
 	}
 }
+
+func AssertResponseWriter(t *testing.T, w http.ResponseWriter, req *http.Request) {
```
assert what?

tkashem replied:
renamed to `AssertResponseWriterImplementsExtendedInterfaces`

tkashem committed:
split audit ResponseWriter construction test

tkashem committed:
improve the ResponseWriter decorator construction test

tkashem committed:
responsewriter decorator tests should use a common test

tkashem committed:
refactor endpoints/responsewriter tests

- move assert logic to the shared testing package
- use "{package name}_test" to avoid circular dependencies

tkashem committed:
refactor endpoints/responsewriter - merge the two tests into one

tkashem committed:
extend the endpoints/responsewriter TestWrapForHTTP1Or2 test

tkashem committed:
add a new test in endpoints/responsewriter

tkashem commented on staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/wrapper_test.go:
```diff
@@ -29,108 +29,68 @@ import (
 	responsewritertesting "k8s.io/apiserver/pkg/endpoints/responsewriter/testing"
 )
 
```
this commit does not introduce any change in the test, it just merges the two tests into one.

tkashem commented:
sharing stress test results here:
```
$ go test -race k8s.io/apiserver/pkg/server -c
$ stress -p=32 ./server.test -test.run=TestWrappedResponseWriterWithFullHandlerChain
5s: 76 runs so far, 0 failures
10s: 188 runs so far, 0 failures
15s: 304 runs so far, 0 failures
...
7m50s: 8499 runs so far, 0 failures
7m55s: 8590 runs so far, 0 failures
8m0s: 8682 runs so far, 0 failures
```

tkashem commented:
@sttts @wojtek-t I have split the commits to make review easier, and shared the `stress` test result above
